src/main/java/org/broadinstitute/hellbender/engine/ReadsPathDataSource.java

```diff
@@ -0,0 +1,520 @@
+package org.broadinstitute.hellbender.engine;
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.MergingSamRecordIterator;
+import htsjdk.samtools.SAMException;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.SamFileHeaderMerger;
+import htsjdk.samtools.SamInputResource;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.IOUtil;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.iterators.SAMRecordToReadIterator;
+import org.broadinstitute.hellbender.utils.iterators.SamReaderQueryingIterator;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadConstants;
+
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Manages traversals and queries over sources of reads which are accessible via {@link Path}s
+ * (for now, SAM/BAM/CRAM files only).
+ *
+ * Two basic operations are available:
+ *
+ * -Iteration over all reads, optionally restricted to reads that overlap a set of intervals
+ * -Targeted queries by one interval at a time
+ */
+public final class ReadsPathDataSource implements ReadsDataSource {
```

Comment:
The way you've structured this PR causes the commit history for `ReadsDataSource` to be lost/obfuscated (I checked just now in an IDE), which we should avoid if possible. I think we can help git understand what's happening by using a different name for the interface in this PR, merge it, then open a second PR to rename the interface to `ReadsDataSource`.

Reply:
I think it has to be two PRs unfortunately, as with two commits I think it'd just get confused again after the squash. If you keep the rename of `ReadsDataSource` -> `ReadsPathDataSource` in this PR, but name the interface something like `ReadsDataSourceInterface` temporarily git should be able to follow the changes (but you should test this). If it works, we can do a second PR to rename the interface after this is merged.

Reply:
We can just not squash them on merge.